Introduce Site Review to the Onboarding Wizard

delawski opened the pull request:
## Summary

The main update this PR brings is adding a Site Review feature to the last step of the Onboarding Wizard:

<img width="1261" alt="Screenshot 2021-09-08 at 15 23 30" src="https://user-images.githubusercontent.com/478735/132517573-9d037f92-0001-4f50-a1f9-0503db149871.png">

Along with the Site Review, a new panel has been added to the AMP Settings screen. It can be dismissed by a user (this gets stored in the user meta) and re-appears whenever a template mode is changed (after saving settings).

![Screenshot 2021-09-08 at 15 24 33](https://user-images.githubusercontent.com/478735/132517750-2eb7cfda-e30b-4e87-860b-df8c0abce37e.png)

Besides adding the Site Review feature, this PR updates some of the microcopy and styles in the Onboarding Wizard (which partially covers #4719).

| Welcome Screen | Technical Background | Template Mode |
| --- | --- | --- |
| <img width="1274" alt="Screenshot 2021-09-08 at 15 27 27" src="https://user-images.githubusercontent.com/478735/132518449-31da0236-2aea-480e-8801-2d2ee80f5de0.png"> | <img width="1261" alt="Screenshot 2021-09-08 at 15 27 33" src="https://user-images.githubusercontent.com/478735/132518470-0a9f2b3e-2377-47f2-8f14-9a2a736bbff8.png"> | <img width="1264" alt="Screenshot 2021-09-08 at 15 27 39" src="https://user-images.githubusercontent.com/478735/132518461-741ca942-19f5-408b-a9a0-7acdadb3184e.png"> |

Note that the target branch for this PR is `feature/5578-add-new-settings` (#6501).

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6071

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Update "Welcome" step copy

delawski committed:
Update plugin name copy and color

delawski committed:
Determine active page index based on current page kept in state

The list of Onboarding Wizard pages can be dynamically changed depending on the theme mode. The actual pages list is memoized in the `adaptedPages` array. The `adaptedPages` list should be used as a source of truth (after all, the `activePageIndex` references the `adaptedPages` array).

In some cases the `activePageIndex` can get out of sync with the `adaptedPages` array. For instance, when viewing the last page and one of the previous pages gets removed the `activePageIndex` references a non-existent item.

With this commit, we're stop keeping active page index in the state and instead track current page itself. Based on the `currentPage` object and the `adaptedPages` array we can determine the `activePageIndex`. This eliminates a risk of referencing a non-existent array item.

delawski committed:
Combine Summary and Done steps of Wizard into single step

delawski committed:
Use uniform name for Review (final) step in Onboarding Wizard

delawski committed:
Remove unused code

delawski committed:
Ensure saving message is displayed as long as there are unsaved options

delawski committed:
Use grid layout for review step and update content hierarchy

delawski committed:
Add preview page and mode selector to Review step

delawski committed:
Extract preview to separate component; add loading state

delawski committed:
Refactor `RadioGroup` into stand-alone component

delawski committed:
Improve Review site Onboarding step styles

delawski committed:
Add links to "Need help" section in "Review" step

delawski committed:
Update "Review" step E2E test specs

delawski committed:
Update copy and typography

delawski committed:
Replace radio buttons with toggle for preview mode

delawski committed:
Update `RadioGroup` typography

delawski committed:
Update E2E tests

delawski committed:
Add "Review" section to Settings page

delawski committed:
Update unit tests

delawski committed:
Use more specific name for `SiteReview` component

delawski committed:
Add "Browse Site" button to AMP Settings page

delawski committed:
Prevent nesting HTML headings

delawski committed:
Add E2E test for Review section on Settings page

delawski committed:
Add basic dismiss functionality to site Review section

delawski committed:
Add user meta and REST field for Review panel `is_dismissed` state

delawski committed:
Remove rebase artifact

delawski committed:
Use Review panel dismiss state from user meta on Settings screen

delawski committed:
Move "Review" panel dismissal logic to client-side

delawski committed:
Add E2E tests for "Review" panel on AMP settings page

delawski committed:
Make Review panel dismissal state property optional

delawski committed:
Clean up and add unit tests

delawski committed:
Remove unused components

delawski committed:
Update copy and styles

delawski committed:
Rename "Review" step to "Done" in OW and update copy

delawski committed:
Fix E2E tests

delawski committed:
Ensure stroke width is proportional to icon size

delawski committed:
Convert `RadioGroup` into `NavMenu`

delawski committed:
Update snapshots

delawski committed:
Update copy in Review panel on AMP Settings screen

delawski committed:
Use constants for checking theme support / template mode

delawski committed:
Fix PHPCS issues

delawski committed:
Fix stylelint issues

delawski committed:
Move User REST extension to separate file; add tests

delawski committed:
Ensure site Review panel shows up before proceeding

delawski committed:
Fix phpcs issues

delawski commented:
~(Note that this will get rebased once #6501 is approved and merged into `develop`.)~

I've rebased and resolved conflicts already since this work is needed for the Site Scan feature (#4719).

jwold reviewed (approved):
Looks great! Good work.

delawski committed:
Try fixing flaky E2E test

westonruter committed:
Delete amp_review_panel_dismissed_for_template_mode usermeta at uninstall

westonruter committed:
Let UserRESTEndpointExtension be Delayed

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,121 @@
+				'update_callback' => [ $this, 'update_review_panel_dismissed_for_template_mode' ],
+				'schema'          => [
+					'description' => __( 'For which template mode the Review panel on the Settings screen was dismissed by a user', 'amp' ),
+					'type'        => [ 'string', 'bool' ],
```
Why not `enum`?

delawski replied:
I wasn't aware of the `enum` property.

In 680b746 I've switched to `enum` and also removed a way to clear the flag value since we haven't used it anywhere.

westonruter commented on assets/src/onboarding-wizard/pages/done/index.js:
```diff
@@ -0,0 +1,194 @@
+					<li dangerouslySetInnerHTML={ {
+						__html: sprintf(
+							/* translators: placeholder is a link to the plugin site. */
+							__( '<a href="%s" target="_blank" rel="noreferrer noopener">Learn more</a> how the PX plugin works', 'amp' ),
```
Since we haven't gotten to that point yet:

```suggestion
							__( '<a href="%s" target="_blank" rel="noreferrer noopener">Learn more</a> how the AMP plugin works', 'amp' ),
```

westonruter commented on assets/src/onboarding-wizard/pages/technical-background/index.js:
```diff
@@ -48,47 +48,11 @@ export function TechnicalBackground() {
-				</svg>
 				<h1>
-					{ __( 'Are you technical?', 'amp' ) }
+					{ __( 'Technical background', 'amp' ) }
```
```suggestion
					{ __( 'Technical Background', 'amp' ) }
```

westonruter commented on assets/src/onboarding-wizard/pages/technical-background/index.js:
```diff
@@ -106,13 +70,11 @@ export function TechnicalBackground() {
+							</h2>
 							<p>
-								{ __( 'I am a “Developer or technically savvy” user. I am familiar with HTML, CSS, JavaScript, and PHP. I am able to do WordPress development, including making changes to themes and plugins, as well as assembling full WordPress sites out of plugins and theme components. I can understand and address AMP validation issues.', 'amp' ) }
+								{ __( 'I can do WordPress development such as making changes to themes and plugins; I can build full WordPress sites out of plugins and themes and can address configuration issues and understand.', 'amp' ) }
```
```suggestion
								{ __( 'I can do WordPress development by modifying themes and plugins. I am familiar with PHP, JavaScript, HTML, and CSS.', 'amp' ) }
```

westonruter commented on assets/src/onboarding-wizard/pages/technical-background/index.js:
```diff
@@ -132,13 +94,11 @@ export function TechnicalBackground() {
-								</h2>
-							</div>
+							<h2>
+								{ __( 'Non-technical or Wanting a simpler setup', 'amp' ) }
```
```suggestion
								{ __( 'Non-technical or wanting a simpler setup', 'amp' ) }
```

westonruter commented on assets/src/onboarding-wizard/pages/technical-background/index.js:
```diff
@@ -106,13 +70,11 @@ export function TechnicalBackground() {
-								</h2>
-							</div>
+							<h2>
+								{ __( 'Developer or Technically Savvy', 'amp' ) }
```
```suggestion
								{ __( 'Developer or technically savvy', 'amp' ) }
```

westonruter commented on assets/src/settings-page/site-review.js:
```diff
@@ -0,0 +1,127 @@
+					<li dangerouslySetInnerHTML={ {
+						__html: sprintf(
+							/* translators: placeholder is a link to the plugin site. */
+							__( '<a href="%s" target="_blank" rel="noreferrer noopener">Learn more</a> how the PX plugin works', 'amp' ),
```
```suggestion
							__( '<a href="%s" target="_blank" rel="noreferrer noopener">Learn more</a> how the AMP plugin works', 'amp' ),
```

westonruter commented on assets/src/components/user-context-provider/index.js:
```diff
@@ -125,6 +138,39 @@ export function UserContextProvider( { children, onlyFetchIfPluginIsConfigured =
+			return;
+		}
+
+		setSavingReviewPanelDismissedForTemplateMode( false );
```
Should this go inside the `catch` block?

delawski replied:
Would it make a difference?

If the request succeeds, we'll get out of the `try` block and reach this point. This way the internal saving state is cleared as expected.

If the request fails, the async error is triggered in the `catch` block. It in turn causes a take-over error screen/message to be rendered and the entire React app "dies". So we'll never reach this line (it becomes irrelevant anyway).

westonruter replied:
Ah, ok. 👍

westonruter commented on assets/src/onboarding-wizard/pages/done/index.js:
```diff
@@ -0,0 +1,194 @@
+				</p>
+				{ STANDARD === themeSupport && (
+					<p>
+						{ __( 'In Standard mode there is a single AMP version of your site. Browse your site below to ensure it meets your expectations.', 'amp' ) }
```
Since on desktop the preview is to the side and not below.

```suggestion
						{ __( 'In Standard mode there is a single AMP version of your site. Browse your site here to ensure it meets your expectations.', 'amp' ) }
```

westonruter commented on assets/src/onboarding-wizard/pages/done/index.js:
```diff
@@ -0,0 +1,194 @@
+							{ __( 'In Transitional mode AMP and non-AMP versions of your site are served using your currently active theme.', 'amp' ) }
+						</p>
+						<p>
+							{ __( 'Browse your site below to ensure it meets your expectations, and toggle the AMP setting to compare both versions.', 'amp' ) }
```
```suggestion
							{ __( 'Browse your site here to ensure it meets your expectations, and toggle the AMP setting to compare both versions.', 'amp' ) }
```

westonruter commented on assets/src/onboarding-wizard/pages/done/index.js:
```diff
@@ -0,0 +1,194 @@
+				{ READER === themeSupport && (
+					<>
+						<p>
+							{ __( 'In Reader mode AMP is served using your selected Reader theme, and pages for your non-AMP site are served using your primary theme. Browse your site below to ensure it meets your expectations, and toggle the AMP setting to compare both versions.', 'amp' ) }
```
```suggestion
							{ __( 'In Reader mode AMP is served using your selected Reader theme, and pages for your non-AMP site are served using your primary theme. Browse your site here to ensure it meets your expectations, and toggle the AMP setting to compare both versions.', 'amp' ) }
```

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,130 @@
+				__( 'Sorry, the template mode is incorrect.', 'amp' ),
+				[ 'status' => 400 ]
+			);
+		}
```
This wouldn't be needed if using an `enum`.

delawski replied:
Check out 680b746.

westonruter commented on assets/src/onboarding-wizard/pages/done/use-preview.js:
```diff
@@ -0,0 +1,71 @@
+			return __( 'Search results', 'amp' );
+
+		default:
+			return `${ type.charAt( 0 ).toUpperCase() }${ type.slice( 1 ) }`;
```
This isn't working so well. I see `Post_tag`: 

![image](https://user-images.githubusercontent.com/134745/135025540-8c696dba-0e92-4c03-9fc1-a7478037c82f.png)

Also, this won't be showing up in the user's locale. 

To fix this properly, I think the `\AmpProject\AmpWP\Validation\ScannableURLProvider::get_urls()` method should return a `label` (or `title`) in addition to the `url` and `type` for each entry. Then there would not be a need for the `getTitle` here.

delawski replied:
This has been addressed in eca7274.

Note that we may still have issues with the custom taxonomy and the custom post type labels. For instance, WooCommerce registers a `product_cat` taxonomy. Its `name` is "Product categories" and its `singular_name` is just "Category". The core Category `name` is "Categories" and the `singular_name` is "Category", too.

I went with the `name` label since it best differentiates the taxonomies. The drawback is that it is usually a plural noun and probably should be used standalone (unless we're okay with having "Categories Archive" or "Tags Archive").

westonruter commented on src/Admin/OnboardingWizardSubmenuPage.php:
```diff
@@ -237,8 +247,9 @@ public function enqueue_assets( $hook_suffix ) {
-			'FINISH_LINK'                        => $amp_settings_link,
+			'SETTINGS_LINK'                      => $amp_settings_link,
 			'OPTIONS_REST_PATH'                  => '/amp/v1/options',
+			'PREVIEW_URLS'                       => $this->get_preview_urls( $this->scannable_url_provider->get_urls() ),
```
See note above about how `get_urls` should return labels in addition to `url` and `type`.

westonruter commented on assets/src/onboarding-wizard/pages/done/index.js:
```diff
@@ -0,0 +1,194 @@
+						{ __( 'There was an error downloading your Reader theme. As a result, your site is currently using the legacy reader theme. Please install your chosen theme manually.', 'amp' ) }
+					</AMPNotice>
+				) }
+				{ STANDARD !== themeSupport && (
+					<AMPSettingToggle
+						text={ __( 'AMP', 'amp' ) }
+						checked={ isPreviewingAMP }
+						onChange={ toggleIsPreviewingAMP }
+						compact={ true }
+					/>
+				) }
```
I noticed that on non-AMP there is a gap appearing in the page where the admin bar normally is:


https://user-images.githubusercontent.com/134745/135026552-d3ffa06d-3bad-4a0b-8173-fb9156f77931.mov

The issue seems to to be that the conditional in the following code is not allowing the `script` to be added to non-AMP pages:

https://github.com/ampproject/amp-wp/blob/73cfc46fe212d7dbb052121a995d2cb15e156322/includes/amp-helper-functions.php#L183-L219

It seems what is needed to fix this is to make the selector more specific:

```diff
diff --git a/includes/amp-helper-functions.php b/includes/amp-helper-functions.php
index d7fbc6f0d..8b55fc218 100644
--- a/includes/amp-helper-functions.php
+++ b/includes/amp-helper-functions.php
@@ -202,7 +202,7 @@ function() {
 					/** @type {HTMLStyleElement} */
 					const style = document.createElement( 'style' );
 					style.setAttribute( 'type', 'text/css' );
-					style.appendChild( document.createTextNode( 'html { margin-top: 0 !important; } #wpadminbar { display: none !important; }' ) );
+					style.appendChild( document.createTextNode( 'html:not(#_) { margin-top: 0 !important; } #wpadminbar { display: none !important; }' ) );
 					document.head.appendChild( style );
 
 					document.addEventListener( 'DOMContentLoaded', function() {
```

Otherwise, it is getting overridden:

![image](https://user-images.githubusercontent.com/134745/135027102-c79fbc30-150f-47a1-bfef-e4d5889d6a42.png)

Specifically by this style:

```html
<style media="screen">
	html { margin-top: 32px !important; }
	* html body { margin-top: 32px !important; }
	@media screen and ( max-width: 782px ) {
		html { margin-top: 46px !important; }
		* html body { margin-top: 46px !important; }
	}
</style>
```

Either that, or more simply the logic could be run at `wp_head` with priority 20, to go above the priority 10 at which `_admin_bar_bump_cb()` is outputting the above styles:

```diff
diff --git a/includes/amp-helper-functions.php b/includes/amp-helper-functions.php
index d7fbc6f0d..8b0235e23 100644
--- a/includes/amp-helper-functions.php
+++ b/includes/amp-helper-functions.php
@@ -187,7 +187,7 @@ static function() {
 	 * by Preview component located in <assets/src/setup/pages/save/index.js>.
 	 */
 	add_action(
-		'wp_print_scripts',
+		'wp_head',
 		function() {
 			if ( ! amp_is_dev_mode() || ! is_admin_bar_showing() ) {
 				return;
@@ -215,7 +215,8 @@ function() {
 				} )();
 			</script>
 			<?php
-		}
+		},
+		20
 	);
 }
```

We'd just have to make sure it also runs at `amp_post_template_head` I suppose.

delawski replied:
I really like the simplicity of the first approach with bumping up the selector specificity. I've implemented it in 28b4583.

delawski committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Bump selector specificity so that no space is reserved for admin bar

delawski committed:
Use `enum` for review panel dismissed for template mode REST param values

delawski committed:
Determine preview pages labels on backend

delawski committed:
Bring back ability to clear review panel dismissed for template mode flag

delawski committed:
Update E2E tests

delawski committed:
Update unit test

delawski committed:
Update unit test

delawski committed:
Fix PHPCS issues

delawski commented:
@westonruter I've addressed your feedback. Unfortunately, there's (again) an issue with a failing E2E test. I can look into that in several hours (tomorrow morning).

delawski commented:
> Unfortunately, there's (again) an issue with a failing E2E test. I can look into that in several hours (tomorrow morning).

Argh! It passed on a second try.

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -57,7 +57,13 @@ public function register_rest_field() {
-					'type'        => [ 'string', 'bool' ],
+					'type'        => 'string',
+					'enum'        => [
+						null,
```
Is `null` still valid here? Given the `type` is `string`.

westonruter replied:
Aside: I don't believe that `null` can be stored in in user meta. Storing `null` results in an empty string being returned. Similarly, the `get_review_panel_dismissed_for_template_mode` method says it returns a `string` or `WP_Error`, not `null`.

westonruter replied:
So should this rather be:

```suggestion
						'',
```

delawski replied:
This was a bit of a puzzle to me. Another approach I considered was to set

```php
'type' => [ 'string', 'null' ]
```

but then the field validation didn't recognize `null` as an allowed `enum` value.

> I don't believe that null can be stored in in user meta.

What we're actually doing if the value is `empty` is deleting the user meta.

Anyway, let's go with an empty string (which we eventually received in `update_review_panel_dismissed_for_template_mode` even if `null` was used here).

westonruter commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -100,15 +106,17 @@ public function get_urls( $offset = 0 ) {
-				'type' => 'search',
+				'url'   => $url,
+				'type'  => 'search',
+				'label' => __( 'Search Results Page', 'amp' ),
```
```suggestion
				'label' => __( 'Search Results', 'amp' ),
```

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,122 @@
+	 *
+	 * @param array $user Array of user data prepared for REST.
+	 *
+	 * @return string|WP_Error Template mode fir which the panel is dismissed, empty string if the option has not been set, or WP_Error if the current user lacks permission.
```
```suggestion
	 * @return string|WP_Error Template mode for which the panel is dismissed, empty string if the option has not been set, or WP_Error if the current user lacks permission.
```

westonruter commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -69,9 +70,11 @@ public function get_urls( $offset = 0 ) {
+					$urls[]           = [
+						'url'   => get_permalink( $post_ids[0] ),
+						'type'  => $post_type,
+						'label' => empty( $post_type_object->labels->name ) ? $post_type : $post_type_object->labels->name,
```
Following up on what you said in https://github.com/ampproject/amp-wp/pull/6596#discussion_r717788251:

> Note that we may still have issues with the custom taxonomy and the custom post type labels. For instance, WooCommerce registers a `product_cat` taxonomy. Its `name` is "Product categories" and its `singular_name` is just "Category". The core Category `name` is "Categories" and the `singular_name` is "Category", too.
>
> I went with the `name` label since it best differentiates the taxonomies. The drawback is that it is usually a plural noun and probably should be used standalone (unless we're okay with having "Categories Archive" or "Tags Archive").

I think that WooCommerce may be in error here to not have chosen “Product category” as the singular name. I see they have also used “Categories” as the `label` too: https://github.com/woocommerce/woocommerce/blob/55ae972e8abf5ef5a53bbf6338f433f8927bff84/includes/class-wc-post-types.php#L93-L109

Therefore, I think we can just go ahead and use the `singular_name`.

```suggestion
						'label' => $post_type_object->labels->singular_name ?: $post_type,
```

westonruter commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -80,18 +83,21 @@ public function get_urls( $offset = 0 ) {
+					$urls[]          = [
+						'url'   => $link,
+						'type'  => $taxonomy,
+						'label' => empty( $taxonomy_object->labels->name ) ? $taxonomy : $taxonomy_object->labels->name,
```
```suggestion
						'label' => $taxonomy_object->labels->singular_name ?: $taxonomy,
```

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,122 @@
+	/**
+	 * Updates a user's setting determining for which template mode the "Review" panel was dismissed.
+	 *
+	 * @param string  $template_mode Template mode.
```
```suggestion
	 * @param string  $template_mode Template mode or empty string.
```

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,122 @@
+			'user',
+			self::USER_FIELD_REVIEW_PANEL_DISMISSED_FOR_TEMPLATE_MODE,
+			[
+				'get_callback'    => [ $this, 'get_review_panel_dismissed_for_template_mode' ],
+				'update_callback' => [ $this, 'update_review_panel_dismissed_for_template_mode' ],
```
At first I thought this was missing a `permission_callback`, but I see that is only relevant for routes.

delawski commented on tests/php/src/Admin/UserRESTEndpointExtensionTest.php:
```diff
@@ -0,0 +1,106 @@
+		$this->assertWPError( $this->user_rest_endpoint_extension->update_review_panel_dismissed_for_template_mode( AMP_Theme_Support::TRANSITIONAL_MODE_SLUG, $editor_user ) );
+
+		$this->assertEquals( AMP_Theme_Support::TRANSITIONAL_MODE_SLUG, get_user_meta( $admin_user->ID, UserRESTEndpointExtension::USER_FIELD_REVIEW_PANEL_DISMISSED_FOR_TEMPLATE_MODE, true ) );
+		$this->assertTrue( $this->user_rest_endpoint_extension->update_review_panel_dismissed_for_template_mode( null, $admin_user ) );
```
```suggestion
		$this->assertTrue( $this->user_rest_endpoint_extension->update_review_panel_dismissed_for_template_mode( '', $admin_user ) );
```

delawski commented on tests/php/src/Admin/OnboardingWizardSubmenuPageTest.php:
```diff
@@ -167,4 +167,55 @@ public function test_get_close_link( $referrer_link_callback, $expected_referrer
+			[
+				'type'  => 'search',
+				'url'   => 'https://example.com/?s=foobar',
+				'label' => 'Search Results Page',
```
```suggestion
				'label' => 'Search Results',
```

delawski commented on tests/php/src/Admin/OnboardingWizardSubmenuPageTest.php:
```diff
@@ -167,4 +167,55 @@ public function test_get_close_link( $referrer_link_callback, $expected_referrer
+				'type'    => 'search',
+				'url'     => 'https://example.com/?s=foobar',
+				'amp_url' => amp_add_paired_endpoint( 'https://example.com/?s=foobar' ),
+				'label'   => 'Search Results Page',
```
```suggestion
				'label'   => 'Search Results',
```

delawski commented on tests/e2e/utils/onboarding-wizard-utils.js:
```diff
@@ -132,7 +124,10 @@ export async function cleanUpSettings() {
-			wp.apiFetch( { path: '/wp/v2/users/me', method: 'POST', data: { amp_dev_tools_enabled: true } } ),
+			wp.apiFetch( { path: '/wp/v2/users/me', method: 'POST', data: {
+				amp_dev_tools_enabled: true,
+				amp_review_panel_dismissed_for_template_mode: null,
```
```suggestion
				amp_review_panel_dismissed_for_template_mode: '',
```

delawski committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Unify code style

delawski committed:
Fix E2E test

delawski commented:
@westonruter I've addressed your feedback and also _hopefully_ (I couldn't replicate the error locally) fixed that flaky E2E test.

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,122 @@
+	 */
+	public function get_review_panel_dismissed_for_template_mode( $user ) {
+		if ( wp_get_current_user()->ID !== $user['id'] ) {
+			return new WP_Error(
+				'amp_rest_cannot_get_other_user',
+				__( 'Sorry, the current user is not allowed to get this data about other user.', 'amp' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
```
I just realized that this should probably:

```suggestion
			return null;
```

Otherwise, when using the REST API to fetch users, then this field shows up as an error:

> ![image](https://user-images.githubusercontent.com/134745/135313968-e6ce7d5c-e773-4e0b-9e17-dab256f07ab4.png)

I'll need to double-check the best practice for visibility of a field like this.

westonruter replied:
That being said, we currently allow the `amp_dev_tools_enabled` field to be exposed to other users, so perhaps this should be visible by default too?

westonruter replied:
Only `administrator` users are allowed to `list_users` for the `/users` endpoint, and for `/users/:id` a request to non-self also requires `list_users` cap. So it's not a problem to expose this information.

westonruter replied:
Updating this now…

westonruter committed:
Remove extraneous cap checks and update tests to use request dispatching

westonruter commented on src/Admin/UserRESTEndpointExtension.php:
```diff
@@ -0,0 +1,106 @@
+	 * @return bool|WP_Error The result of update_user_meta, or WP_Error if the current user lacks permission.
+	 */
+	public function update_review_panel_dismissed_for_template_mode( $template_mode, WP_User $user ) {
+		if ( ! current_user_can( 'edit_user', $user->ID ) ) {
+			return new WP_Error(
+				'amp_rest_cannot_edit_user',
+				__( 'Sorry, the current user is not allowed to make this change.', 'amp' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
```
I also realized that this is not needed either, because the `update_callback` will not run for the user endpoint if the user cannot `edit_user` thus the Users controller will return false when `update_item_permissions_check` is called.

westonruter committed:
Add test assertion for enum

westonruter committed:
Fix phpdoc for get_review_panel_dismissed_for_template_mode return value

westonruter committed:
Remove extraneous edit_user cap check

westonruter committed:
Fix phpdoc for update_review_panel_dismissed_for_template_mode after a1f76832

westonruter committed:
Verify that editor user cannot access another user

westonruter committed:
Remove unused PHP import

westonruter reviewed (approved):
Great work!

delawski commented:
@westonruter Thanks for removing the redundant cap checks, improving the tests and pushing the PR through the finish line!
